backend/src/services/user-secret/user-secret-service.ts

```diff
@@ -0,0 +1,210 @@
+    id: secret.id,
+    name: secret.name,
+    type: secret.type as UserSecretType,
+    data,
+    createdAt: secret.createdAt.toISOString(),
+    updatedAt: secret.updatedAt.toISOString(),
+    createdBy: secret.createdBy
+  });
+
+  const listUserSecrets = async ({
+    actor,
+    actorId,
+    actorAuthMethod,
+    actorOrgId,
+    offset,
+    limit
+  }: TListUserSecretsDTO): Promise<TListUserSecretsResponse> => {
+    await validatePermission({ actor, actorId, actorAuthMethod, actorOrgId }, OrgPermissionActions.Read);
+
+    const { secrets, totalCount } = await userSecretDAL.findUserSecrets(actorOrgId, { offset, limit });
+
+    const decryptedSecrets = await Promise.all(
+      secrets.map(async (secret) => {
+        if (!secret.encryptedData) {
+          return null;
+        }
+
+        logger.info({ secret }, `Decrypting secret`);
+
+        const decryptedData = decryptSecretData(secret.encryptedData);
+        return formatSecretResponse(secret, JSON.parse(decryptedData) as TUserSecretData);
+      })
+    );
+
+    return {
+      secrets: decryptedSecrets.filter((s): s is TUserSecretResponse => s !== null),
+      totalCount
+    };
+  };
+
+  const getUserSecretById = async ({
+    secretId,
+    actor,
+    actorId,
+    actorAuthMethod,
+    actorOrgId
+  }: TGetUserSecretDTO): Promise<TUserSecretResponse> => {
+    await validatePermission({ actor, actorId, actorAuthMethod, actorOrgId }, OrgPermissionActions.Read);
+
+    const secret = await userSecretDAL.findUserSecretById(secretId, actorOrgId);
+    if (!secret) {
+      throw new NotFoundError({ message: "User secret not found" });
+    }
+
+    if (!secret.encryptedData) {
+      throw new NotFoundError({ message: "Secret data not found" });
+    }
+
+    const decryptedData = decryptSecretData(secret.encryptedData);
+
+    return formatSecretResponse(secret, JSON.parse(decryptedData) as TUserSecretData);
+  };
+
+  const createUserSecret = async ({
+    actor,
+    actorId,
+    actorAuthMethod,
+    actorOrgId,
+    name,
+    type,
+    data
+  }: TCreateUserSecretDTO): Promise<TUserSecretResponse> => {
+    await validatePermission({ actor, actorId, actorAuthMethod, actorOrgId }, OrgPermissionActions.Create);
+
+    const encryptedSecret = encryptSecretData(JSON.stringify(data));
+
+    const secret = await userSecretDAL.createUserSecret({
+      name,
+      type,
+      organizationId: actorOrgId,
+      encryptedData: encryptedSecret,
+      createdBy: actorId,
+      iv: "",
+      tag: ""
+    });
+
+    return formatSecretResponse(secret, data);
+  };
+
+  const updateUserSecret = async ({
+    secretId,
+    actor,
+    actorId,
+    actorAuthMethod,
+    actorOrgId,
+    name,
+    data
+  }: TUpdateUserSecretDTO): Promise<TUserSecretResponse> => {
+    await validatePermission({ actor, actorId, actorAuthMethod, actorOrgId }, OrgPermissionActions.Edit);
+
+    // TODO: As transaction
```

Comment:
This update is a read-then-write. It should be wrapped in a transaction so it can be executed atomically.

The right way to do this is to create a "transaction manager" that encapsulates the logic of transaction management, which is then used as a dependency both here and in the DAL. However, I think this is outside the scope of the exercise.

Reply:
I realized after writing this that I can just update conditionally, so no transaction needed.